### PR TITLE
Typecore.type_pat: refactor ppat_of_type and Need_backtrack in wildcards

### PR DESCRIPTION
**Pull request by garrigue**
There is a small, desirable change of behavior in this PR:
we do not backtrack on all wildcard pattern explosions, only on those where `ppat_of_type` knows there will be GADTs.
As a result, some exhaustivity counter-examples have been improved with more cases (see testsuite for an example).

Also use correct names for constructors in `pat_of_constr` and `pats_of_type`.
The dummy name was a legacy of when the name wasn't in the descriptor.

Joint work with @Octachron and @gasche .

### CODE REVIEW

File: typing/parmatch.ml
Comment (trefis) [inline typing/parmatch.ml:999]: I know this code predates the present PR, but can you explain why we do this only when all constructors are GADTs, instead of whenever some constructor is a GADT?

**Review by trefis — APPROVED** [review]

File: typing/parmatch.ml
Comment (gasche) [reply typing/parmatch.ml:999]: In counter-example mode, what `type_pat` checks is that the counter-example pattern is well-typed and inhabited (non-empty). A wildcard is suspicious because it might be non-empty(precisely: there may be no search path where its type is inhabited), so we explode wildcards using `pats_of_type` to check that they contain values, in a best-effort way (we may stop exploding too early).

If there is at least one non-gadt constructor, the type-checker assumes that this constructor is valid in the current context, and non-empty (this is a simplification that makes the check incomplete), so it doesn't try to explode the wildcard. If all constructors are GADTs, then we decide to explode, and (in `type_pat`) we have to backtrack. Effectively this means that the type-checker tries harder on all-GADTs types, by heuristic/choice.

Consider
```
# type empty = |;;
# type 'a singleton = Single of 'a;;
# (function _ -> . : empty singleton -> 'a);;
- : empty singleton -> 'a = <fun>
# (function _ -> . : empty option -> 'a);;
Error: This match case could not be refuted.
       Here is an example of a value that would reach it: _
# (function _ -> . : (empty, empty) result -> 'a);;
Error: This match case could not be refuted.
       Here is an example of a value that would reach it: _
```

In the `empty option` case, this reasoning of the type-checker is correct (the `None` case is always inhabited). The wildcard is not split and is shown as an obscure but valid counter-example.

In the `(empty, empty) result` case, the reasoning of the type-checker is an incomplete approximation: it proposes the unexploded wildcard as a counter-example, and the user would have to explode manually to reject it.

Comment (trefis) [reply typing/parmatch.ml:999]: Thanks for the explanation.

**Comment by trefis** [comment]
(Note: I'll let you guys rebase and merge this whenever convenient)

**Comment by gasche** [comment]
(I just rebased and will merge when the CI goes green again.)

**Pull request merged**
